tests/code/test_spreadsheet.py

```diff
@@ -0,0 +1,194 @@
+from app.models.term import Term
+from app.models.eventParticipant import EventParticipant
+
+from app.logic.spreadsheet import *
+
+
+@pytest.mark.integration
+def test_create_spreadsheet():
+    create_spreadsheet("2020-2021")
+
+@pytest.mark.integration
+def test_calculateRetentionRate():
+    fall_dict=({'Adopt-a-Grandparent': ['khatts'], 'CELTS-Sponsored Event': [None]})
+    spring_dict=({'Hunger Initiatives': ['neillz', 'khatts', 'ayisie', 'partont']})
+    assert calculateRetentionRate(fall_dict,spring_dict) == {'Adopt-a-Grandparent': 0.0, 'CELTS-Sponsored Event': 0.0}
+
+    fall_dict=({'Hunger Initiatives': ['neillz', 'khatts', 'ayisie', 'partont']})
+    assert calculateRetentionRate(fall_dict,spring_dict) == {'Hunger Initiatives': 1.0}
+
+    fall_dict=({'Hunger Initiatives': ['neillz', 'khatts', 'ayisie', 'partont']})
+    spring_dict=({'Hunger Initiatives': ['neillz', 'khatts', 'ayisie']})
+    assert calculateRetentionRate(fall_dict,spring_dict) == {'Hunger Initiatives': 0.75}
+
+@pytest.mark.integration
+def test_removeNullParticipants():
+    test_input_list = ['khatts']
+    assert removeNullParticipants(test_input_list) == ['khatts']
+    test_input_list = ['khatts', '', 'ayisie']
+    assert removeNullParticipants(test_input_list) == ['khatts', 'ayisie']
+
+@pytest.mark.integration
+def test_termParticipation():
+    with mainDB.atomic() as transaction:    
+        assert termParticipation('Fall 2020')=={'Adopt-a-Grandparent': ['khatts'], 'CELTS-Sponsored Event': [None]}
+
+        EventParticipant.create(user = 'partont',
+                                event = 10,
+                                hoursEarned = 1)
+        assert termParticipation('Fall 2020')=={'Adopt-a-Grandparent': ['khatts', 'partont'], 'CELTS-Sponsored Event': [None]}
+
+        EventParticipant.create(user = 'ayisie',
+                                event = 14,
+                                hoursEarned = 1)
+        assert termParticipation('Fall 2020')=={'Adopt-a-Grandparent': ['khatts', 'partont'], 'CELTS-Sponsored Event': ['ayisie']}
+        transaction.rollback()
+
+@pytest.mark.integration
+def test_getRetentionRate():
+    with mainDB.atomic() as transaction:
+        assert getRetentionRate("2020-2021") == [('Adopt-a-Grandparent', '0.0%'), ('CELTS-Sponsored Event', '0.0%')]
+        
+        testEvent = Event.create(name="Test Event",
+                                 term=2,
+                                 program=3)
+        EventParticipant.create(user = 'khatts',
+                                event = testEvent,
+                                hoursEarned = 1)
+        assert getRetentionRate("2020-2021") == [('Adopt-a-Grandparent', '100.0%'), ('CELTS-Sponsored Event', '0.0%')]
+        transaction.rollback()
+
+@pytest.mark.integration
+def test_repeatVolunteers():
+    with mainDB.atomic() as transaction:
+
+        assert list(repeatVolunteers().execute()) == [('Sreynit Khatt', 4), ('Zach Neill', 2)]
```

Comment:
I think like we may have talked about, and like my previous comment which is not necessarily mandatory, but it makes it more clear what the test is doing if we add someone who has not participated in any events to a single event, and verify they do not show up in our first assertion. Then we can add them to another event and verify they do in fact show up in our second. Again, just a little hesitant on relying on test data.

Reply:
An alternative could be to add a comment stating that Tyler Parton belongs to an event already. I would recommend just either make these more explicit by defining things that don't exist, or add comments to make sure people know what we are doing, or even both.